samza-sql/src/main/java/org/apache/samza/sql/translator/JoinTranslator.java

```diff
@@ -0,0 +1,319 @@
+    RexInputRef leftRef = (RexInputRef) operands.get(0);
+    RexInputRef rightRef = (RexInputRef) operands.get(1);
+    if (leftRef.getIndex() > rightRef.getIndex()) {
+      RexInputRef tmpRef = leftRef;
+      leftRef = rightRef;
+      rightRef = tmpRef;
+    }
+
+    // Get the table index and stream index
+    int deltaIdx = rightRef.getIndex() - join.getLeft().getRowType().getFieldCount();
+    streamIds.add(isTablePosOnRight ? leftRef.getIndex() : deltaIdx);
+    tableIds.add(isTablePosOnRight ? deltaIdx : leftRef.getIndex());
+  }
+
+  private String dumpRelPlanForNode(RelNode relNode) {
+    return RelOptUtil.dumpPlan("Rel expression: ",
+        relNode, SqlExplainFormat.TEXT,
+        SqlExplainLevel.EXPPLAN_ATTRIBUTES);
+  }
+
+  private boolean isTable(RelNode relNode) {
+    // NOTE: Any intermediate form of a join is always a stream. Eg: For the second level join of
+    // stream-table-table join, the left side of the join is join output, which we always
+    // assume to be a stream. The intermediate stream won't be an instance of EnumerableTableScan.
+    return relNode instanceof EnumerableTableScan &&
+        sourceResolver.isTable(String.join(".", relNode.getTable().getQualifiedName()));
+  }
+
+  private final class SamzaSqlRelMessageJoinFunction implements StreamTableJoinFunction<SamzaSqlCompositeKey,
+      SamzaSqlRelMessage, KV<SamzaSqlCompositeKey, SamzaSqlRelMessage>, SamzaSqlRelMessage> {
+
+    JoinRelType joinRelType;
+    boolean isTablePosOnRight;
+    List<Integer> streamFieldIds;
+    // Table field names are used in the outer join when the table record is not found.
+    List<String> tableFieldNames;
+
+    SamzaSqlRelMessageJoinFunction(JoinRelType joinRelType, boolean isTablePosOnRight, List<Integer> streamFieldIds,
+        List<String> tableFieldNames) {
+      this.joinRelType = joinRelType;
+      this.isTablePosOnRight = isTablePosOnRight;
+      Validate.isTrue((joinRelType.compareTo(JoinRelType.LEFT) == 0 && isTablePosOnRight) ||
+          (joinRelType.compareTo(JoinRelType.RIGHT) == 0 && !isTablePosOnRight) ||
+          joinRelType.compareTo(JoinRelType.INNER) == 0);
+      this.streamFieldIds = streamFieldIds;
+      this.tableFieldNames = tableFieldNames;
+    }
+
+    @Override
+    public SamzaSqlRelMessage apply(SamzaSqlRelMessage message, KV<SamzaSqlCompositeKey, SamzaSqlRelMessage> record) {
+
+      if (joinRelType.compareTo(JoinRelType.INNER) == 0 && record == null) {
+        log.debug("Record not found for the message with key: " + getMessageKey(message));
+        return null;
+      }
+
+      // The resulting join output should be a SamzaSqlRelMessage containing the fields from both the stream message and
+      // table record. The order of stream message fields and table record fields are dictated by the sql query. The
+      // output should also include the keys from both the stream message and the table record.
+      List<String> outFieldNames = new ArrayList<>();
```

Comment:
Is not clear to me, but if  "outFieldNames" is the same for all result messages, consider moving this computation outside apply().